Split TableTransformer.transform() into 3 phases

zaneselvans opened the pull request:
Within each dataset, there are often a set of standard operations that have to happen at
the beginning and the end of each table's transformation. To allow those operations to
be defined just once per dataset, I create 3 separate phases, defined as abstract
methods in the AbstractTableTransformer class:

 * start_transform()
 * main_transform()
 * finish_transform()

These are called by the (no longer abstract) TableTransformer.transform() method and
should be defined by the dataset specific abstract table transformer. Of course they can
also be overridden by the individual concrete table transformer when necessary.

This commit also adds a decorator @cache_df() that enables the optional caching of the
outputs of the individual in the self._cached_dfs dictionary for forensic / debugging
purposes.

This caching is only done if self.cache_dfs = True. By default it is False.

The transform() method removes all cached dataframes after calling finish_transform()
method unless TableTransformer.clear_cached_dfs is False. It is True by default.

For more information on how decorators work, [I found this post useful](https://realpython.com/primer-on-python-decorators/#decorators-with-arguments)

zaneselvans commented:
@cmgosnell if I do any more work on the row dropper methods before you get back to your computer I'll add them to this PR, but I will try and make sure that each commit is a self-contained piece of functionality so you can look at them individually.

cmgosnell reviewed (changes requested):
This looks good overall. I have some questions and request for more docs/explanations.

I think I would prefer going with `transform_{phase}` as a naming convention for these key `transform` abstract methods. But that is not a hill I will die on.

cmgosnell commented on src/pudl/transform/classes.py:
```diff
@@ -431,6 +431,26 @@ def correct_units(df: pd.DataFrame, params: UnitCorrections) -> pd.DataFrame:
+                    f"{self.table_id.value}: Caching df to {key=} "
+                    f"in {func.__name__}()"
+                )
+                self._cached_dfs[key] = df.copy()
```
why a copied version of the df?

zaneselvans replied:
So that subsequent alterations of the dataframe as it continues to be transformed can't alter the cached copy.

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -136,30 +139,8 @@ def params(self) -> Ferc1TableTransformParams:
-        """
-        ...
-
-    def concat_dbf_xbrl(
```
I started creating a separate `start_transform` method and realized that it was literally going to be a one-line wrapper for the `concat_dbf_xbrl` method, which seemed silly -- in the case of the FERC1 tables, this method **is** the early transform.

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -136,30 +139,8 @@ def params(self) -> Ferc1TableTransformParams:
-        ...
-
-    def concat_dbf_xbrl(
+    @cache_df(key="main")
```
Here `key` is literally the key used to store the dataframe in the internal dictionary. It can be any string. The more different `key`s are specified, the more intermediate dataframes are going to get stored. If you use the same `key` more than once then yes, the dataframe gets overwritten.

They way I'm using it here it's an incrementally updating record of the last state the dataframe had when passed through a decorated function successfully. If need be, you could give every step its own distinct key and save them all separately, but that would use a lot of memory.  It might be useful in the midst of development when you're trying to figure out what step a problem is happening in though.

I kept separate `xbrl` and `dbf` dataframes since they have their own origins and go through different paths before merging into the main transform.

zaneselvans commented on src/pudl/analysis/classify_plants_ferc1.py:
```diff
@@ -531,12 +531,14 @@ def plants_steam_assign_plant_ids(
 
 
-def plants_steam_validate_ids(ferc1_steam_df: pd.DataFrame) -> None:
+def plants_steam_validate_ids(ferc1_steam_df: pd.DataFrame) -> pd.DataFrame:
```
It just seemed like having the no-op function was making for a weird special case.

zaneselvans commented on src/pudl/transform/ferc714.py:
```diff
@@ -357,14 +357,11 @@ def respondent_id(tfr_dfs):
+    df["respondent_name_ferc714"] = df.respondent_name_ferc714.str.strip()
+    df.loc[df.eia_code == 0, "eia_code"] = pd.NA
+    # These excludes fake Test IDs -- not real planning areas
+    df = df[~df.respondent_id_ferc714.isin(BAD_RESPONDENTS)]
```
Blargh, sorry for the confusion. I was working on fixing errors that show up with the new pandas 1.5.0, and in the midst of that updated the versions of some dependencies, and I meant to just pull over the updated `setup.py` and `environment.yml` files but I messed up and merged all the pandas 1.5.0 fixes accidentally 😬

zaneselvans commented on src/pudl/transform/ferc1.py:
```diff
@@ -727,21 +710,18 @@ class PlantsSteamFerc1TableTransformer(Ferc1AbstractTableTransformer):
+            .params.categorize_strings["fuel_type_code_pudl"]
+            .categories.keys()
+        )
+        return (
```
I can put it back. It just always feels weird to me to set a variable only to immediately return it without ever referring to it.

zaneselvans commented on src/pudl/transform/classes.py:
```diff
@@ -431,6 +431,26 @@ def correct_units(df: pd.DataFrame, params: UnitCorrections) -> pd.DataFrame:
 #####################################################################################
 # Abstract Table Transformer classes
 #####################################################################################
+def cache_df(key: str = "main") -> Callable:
```
Okay, I've fleshed this out, let me know if you still have questions.

cmgosnell replied:
these docs are very helpful on the context. i added one question that is more about how it works.

zaneselvans replied:
I've added a link to the primer on decorators that I used to build this, so folks can go get a real explanation if they want it!

zaneselvans commented:
I've responded to your questions/comments and added better documentation of the `cache_df` decorator.

Sorry about confusingly merging in some code changes from the `pandas-1.5` branch accidentally.

I still think I prefer the `start_transform()` and `finish_transform()` since it reads more like English ("What does this method do? It starts the transform. What does that one do? It finishes the transform.") but I've switched it for now.

cmgosnell commented on src/pudl/transform/classes.py:
```diff
@@ -431,6 +431,60 @@ def correct_units(df: pd.DataFrame, params: UnitCorrections) -> pd.DataFrame:
+    def _decorator(func: Callable[..., pd.DataFrame]) -> Callable[..., pd.DataFrame]:
+        @wraps(func)
+        def _wrapper(self: AbstractTableTransformer, *args, **kwargs) -> pd.DataFrame:
+            df = func(self, *args, **kwargs)
```
does a `df = func` because the `func` here is expected to return a dataframe? and the `*args` and `**kwargs` are passed in here and in `_wrapper` but not used in this wrapper because  if they weren't they wouldn't be passed into the actual function or method that this decorator will wrap?

zaneselvans replied:
This is definitely confusing! And also apparently the way that decorators are typically written.

The decorator has to do its own work, based on its own arguments (here, caching a dataframe in a dictionary with a particular key), as well as ensuring that the function it is decorating does its work.

Decorators take a function as an argument and return a function. That returned function is then called by Python with the function that the decorator is applied to as its argument. Here the decorated method is `func`, which can take any inputs (as long as the first one is self, since it's a method) but has to return a single dataframe. The wrapper calls that method, passing in whatever arguments it got originally, and then does the decorator's work of caching the dataframe.

I'm probably not doing a good job of explaining this since I just barely understood it well enough to get it to work.

zaneselvans replied:
I'll add a link to the decorator tutorial I learned from:

https://realpython.com/primer-on-python-decorators/#fancy-decorators

cmgosnell reviewed (commented):
Thanks for the decorator docs. some of that seemed inferrable but its nice to make it clear.

On the naming.. I've tended to use ["reverse notation"](https://gist.github.com/sloria/7001839) for many reasons. while its slightly less colloquial, it clearly groups elements and it is easier to use with tab completion.
